content/technical/open-source-as-a-service.md

```diff
@@ -0,0 +1,155 @@
+Title: Your Service is not Open Source
+Date: 2018-02-06 18:35
+Tags: open source, technical
+Slug: open-source-as-a-service
+Category: Technical
+Summary: Sharing the source code to your SaaS is insufficient. We effectively enable users to contribute to the a running service for it to be Open Source.
+
+Sharing the source code to your SaaS is insufficient to make it
+actually be Open Source. Sounds self-contradictory?
+
+Most services that espouse "Open Source", do so by simply
+[throwing the code over the wall](https://github.com/travis-ci/travis-ci).
+It's better than nothing, but really misses the point that powers
+Open Source: enabling users to make a change to the software
+they're using.
+
+Some other Open Source services, such as
+[GitLab.com](https://about.gitlab.com/community/contribute/development/)
+or [ElasticSearch](https://www.elastic.co/about/free-and-open) do include
+the tools used to operate/deploy their service. Pause for applause 👏👏👏
+
+But thats is also insufficient to actually enable Open Source.
+
+In order for a user of the service to begin contributing, they must
+first deploy and operate their own clone: Something that they explicitly
+chose not to do when they elected to use the software as a service.
+
+Even if the user becomes an operator, and deploys a such clone, it
+is devoid of their own data, devoid of other users, of connections
+to other services, of infrastructure scalability, pipelines, monitoring,
+backups, ... you name it.
+
+In any case, users are certainly not able to iterate on their changes
+to a service against their real world use cases.
+
+What's more, without true Open Source principles, we are missing a
+way to have a community develop with a center of gravity around
+the service itself. This leads to companies
+[forking the service](https://aws.amazon.com/blogs/opensource/stepping-up-for-a-truly-open-source-elasticsearch)
+in a way that typical open source projects with a true community
+are resilient to.
+
+I believe that if we enable contributions to services, we reap
+the advantages of true Open Source. That means enabling users to make a
+change to a running service and experience that change themselves.
+
+You still with me?
+
+Yes, that means opening a pull request against a running service
+and running that change before it's merged.
```

Comment:
Connecting that to your point above -- the *only* way that this is feasible is if your own little shard of that SaaS with your PR operates on a cloned subset of the production data. If any random Joe Developer could submit a PR against gitlab that runs on the production data, what trust could you still place in the gitlab Saas? IMHO, it would be "none".

But again, I think having a playground with empty (or cloned data that belongs to you) is actually preferable, as it gives you the possibility of doing thorough, destructive, and risky tests.

Reply:
I'll cover this in follow up articles and playbook. I believe that your statements are just as true of team contributions.
Many teams only merge code once they have gotten it into production. See the GitHub internal contribution process.

But yes, I'm happy leaving this as a controvesial topic to be discussed further.

Reply:
@martinpitt i think the crucial bit that is mentioned above is that this change would only affect your data / how only you consume the service. 

I think of it like this: 
many SaaS offer an option to try their "beta" features or become part of their "fast" channel, with latest changes not-yet fully tested on everyone else. Other SaaS A/B test features all the time, so it's not like all the users of the service experience the same service anyway. 

Switching to "contributor mode" for an Open Source Service would be very similar to those cases but the features you are beta testing are simply your own.